### ISSUE TITLE
CPU spikes when playing streaming service

### ISSUE THREAD

**Issue by th2j** [LMS-Community/slimserver#776]
Hello

i noticed a strange behavior : 

1 -> when i play local files, the cpu charge is very light, mostly 0 and a few spikes at 3% or 6%
2 -> when i played from a streaming service (qobuz), the charge is higher, with many spikes at 40%, and at the beginning it is at 100%

3 -> but when i played from a streming service (qobuz) with a playlist already set when logitech media server starts, the charge is very light (mostly 0 and a few spikes at 3% or 6%, as case 1)

i don't think the case 2 is normal,

Thanks,

**Comment by th2j** [comment]
Another issue : 
When i add an album to the current playlist, the cpu reach 100% for a few seconds.
This is abnormal. This shouldn't use so much power

**Comment by michaelherger** [comment]
Please share some information about your LMS machine (hardware, OS), and players, configuration, additional plugins you're using etc.

**Comment by th2j** [comment]
Here is some informations : 
LMS 8.3.0
Fedora 34
Squeezelite 1.9.9
Native playback from local file or qobuz (flac)
plugins : additional browse modes, don't stop the music, full text search, last fm audioscrobbler, lastmix, material skin, mysqueezebox services handler, online music library integration, qobuz, radio, radio france, save playlist

**Exemple 1 : play from qobuz with a track in queue when LMS starts -> 9,1% CPU stable in time (column 9) :**
    620 squeeze+ -56   0  173564 153024  15984 S   9,1   1,9   0:08.14 squeezeboxserve                                                                                                                             
    620 squeeze+ -56   0  173564 153024  15984 S   9,1   1,9   0:08.15 squeezeboxserve                                                                                                                             
    620 squeeze+ -56   0  173564 153024  15984 S   9,1   1,9   0:08.16 squeezeboxserve                                                                                                                             
    620 squeeze+ -56   0  173564 153024  15984 S   9,1   1,9   0:08.17 squeezeboxserve                                                                                                                             
    620 squeeze+ -56   0  173564 153024  15984 S   9,1   1,9   0:08.18 squeezeboxserve                                                                                                                             
    620 squeeze+ -56   0  173564 153024  15984 S   9,1   1,9   0:08.19 squeezeboxserve                                                                                                                             
    620 squeeze+ -56   0  173564 153024  15984 S   9,1   1,9   0:08.20 squeezeboxserve                                                                                                                             
    620 squeeze+ -56   0  173564 153024  15984 S   9,1   1,9   0:08.21 squeezeboxserve                                                                                                                             
    620 squeeze+ -56   0  173564 153024  15984 S   9,1   1,9   0:08.22 squeezeboxserve                                                                                                                             
    620 squeeze+ -56   0  173564 153024  15984 S   9,1   1,9   0:08.23 squeezeboxserve                                                                                                                             
    620 squeeze+ -56   0  173564 153024  15984 S   9,1   1,9   0:08.24 squeezeboxserve                                                                                                                             
    620 squeeze+ -56   0  173564 153024  15984 S   9,1   1,9   0:08.25 squeezeboxserve                                                                                                                             
    620 squeeze+ -56   0  173564 153024  15984 S   9,1   1,9   0:08.26 squeezeboxserve                                                                                                                             
    620 squeeze+ -56   0  173564 153024  15984 S   9,1   1,9   0:08.27 squeezeboxserve                      

**Exemple 2 : play from qobuz with a track added in queue after LMS starts -> CPU instable in time with spikes at 60% (column 9) :**
    620 squeeze+ -56   0  178136 156948  16504 S   9,1   2,0   0:12.88 squeezeboxserve                                                                                                                             
    620 squeeze+ -56   0  178136 156948  16504 S  60,0   2,0   0:12.94 squeezeboxserve                                                                                                                             
    620 squeeze+ -56   0  178136 156948  16504 S   9,1   2,0   0:12.95 squeezeboxserve                                                                                                                             
    620 squeeze+ -56   0  178136 156948  16504 S  54,5   2,0   0:13.01 squeezeboxserve                                                                                                                             
    620 squeeze+ -56   0  178136 156948  16504 S   9,1   2,0   0:13.02 squeezeboxserve                                                                                                                             
    620 squeeze+ -56   0  178136 156948  16504 R  45,5   2,0   0:13.07 squeezeboxserve                                                                                                                             
    620 squeeze+ -56   0  178136 156948  16504 S   9,1   2,0   0:13.08 squeezeboxserve                                                                                                                             
    620 squeeze+ -56   0  178136 156948  16504 R  27,3   2,0   0:13.11 squeezeboxserve                                                                                                                             
    620 squeeze+ -56   0  178136 156948  16504 S  36,4   2,0   0:13.15 squeezeboxserve                                                                                                                             
    620 squeeze+ -56   0  178136 156948  16504 S   9,1   2,0   0:13.16 squeezeboxserve                                                                                                                             
    620 squeeze+ -56   0  178136 156948  16504 S  54,5   2,0   0:13.22 squeezeboxserve                                                                                                                             
    620 squeeze+ -56   0  178136 156948  16504 S   9,1   2,0   0:13.23 squeezeboxserve                                                                                                                             
    620 squeeze+ -56   0  178136 156948  16504 R  27,3   2,0   0:13.26 squeezeboxserve                                                                                                                             
    620 squeeze+ -56   0  178136 156948  16504 S  27,3   2,0   0:13.29 squeezeboxserve                                                                                                                             
    620 squeeze+ -56   0  178136 156948  16504 S   9,1   2,0   0:13.30 squeezeboxserve                                                                                                                             
    620 squeeze+ -56   0  178136 156948  16504 R  20,0   2,0   0:13.32 squeezeboxserve

**Comment by michaelherger** [comment]
What hardware?

Can you give step-by-step instruction how to reproduce this? "when LMS starts" doesn't really make sense, because you can't play music without LMS already running.

**Comment by th2j** [comment]
on a nuc NUC6CAYH with celeron J3455 and 8go RAM.
Realtime kernel. And LMS is isolate on CPU 1 with taskset.

**Exemple 2, i start the nuc and LMS, i put a track on queue and play,**
i record the CPU use for about 1 minute with :
     top -d 0,1 -n 10000 > top.txt
and then i do :
     cat /home/hifi/top.txt | grep squeezeboxserve

**Exemple 1, i put a track X on queue, and then restart the PC and LMS. Then i play the track X**

**Comment by mherger** [comment]
So... one of the differences is: in one case the playlist already has been populated, metadata has been loaded etc. In the second case all this needs to be done first - which obviously will use some CPU cycles. 

But there are so many other factors which could come into play: how many control points are active at the time (remote controls, mobile apps, web browsers), what kind of control points? Any other external influences, like automation systems interacting with LMS? Is LMS rescanning in the background? Is there anything particular in your server.log file?

I don't think this is a general issue, but something specific to your installation. Many of us run LMS on much less powerful systems like a Raspberry Pi. My Pi 3 right now shows 0.1% CPU activity for LMS, while playing music...

I'd suggest you join forums.slimdevices.com to discuss the topic. There might be someone who's seen similar behaviour.

**Comment by th2j** [comment]
I don't think this is related to my installation.

I found something right now,
when i play from the plugin qobuz it is ok,
but when i play qobuz music from my database (qobuz music is integrated with the plugin "online music library integration"), the problem happens

**Comment by michaelherger** [comment]
You could try to profile what's going on using `Devel::NYTProf` (https://metacpan.org/pod/Devel::NYTProf):

```
$ perl -d:NYTProf slimserver.pl

$ nytprofhtml -f NYTProf/nytprof.out -o NYTProf/smallartwork --open
```

Don't run LMS too long with NYTProf enabled, as it can generate a ton of data. Just try to run once with a locally stored track, once using the Qobuz menu directly right after server restart. Then compare the resulting diagrams.

**Comment by th2j** [comment]
Thanks

Here is the results :

For playback qobuz flac from database :
https://www.dropbox.com/s/ultw1zlca2flu53/nytprof-1.7z?dl=0

For playback qobuz flac from plugin qobuz :
https://www.dropbox.com/s/mm6j5o8zwa8fcqz/nytprof-2.7z?dl=0

**Comment by mherger** [comment]
That actually was more interesting than I expected it to be 😁 . So the big difference seems to be that for the imported tracks LMS does a lot of DB lookups to get information related to the track (contributors, album) when a controller (web UI, mobile app etc.) requests status information. For the online only tracks this does not happen, as that information is not available there. You can see a big difference around eg. usr-lib-perl5-vendor_perl-Slim-Control-Queries-pm-377-line.html#5041. In the online case LMS would only spend a few ms there, in the "imported" case it's several seconds.

I wonder why the DB seems to be so slow on your system. What's the memory setting you're using (Settings/Advanced/Performance)? It should be "maximum" in your case.

**Comment by th2j** [comment]
yes it is on maximum, and i have a ssd

**Comment by mherger** [comment]
I can't reproduce this issue. Even on the Pi3B the CPU load stays in the low single digit regions. Are you using a particularly large collection? Would you mind sharing your library.db and log files with me?

**Comment by th2j** [comment]
wich log do you need ? there are many i can activate

**Comment by th2j** [comment]
i have 1300 qobuz albums

**Comment by mherger** [comment]
Just the server.log.zip from Settings/Information.

How many local albums?

**Comment by th2j** [comment]
ok but do you need i put log to debug for some logs before, for exemple "player.source" ?

**Comment by th2j** [comment]
5 local albums

**Comment by mherger** [comment]
No additional logging required right now. I only want to see whether there's something obvious.

**Comment by th2j** [comment]
there is nothing, just the line "Starting Logitech Media Server"

**Comment by th2j** [comment]
here is the library.db
https://www.dropbox.com/s/9mdgpk08fdzfhew/library.db?dl=0

**Comment by th2j** [comment]
New problem that happens today with qobuz :
the sample rate has fallen to 44,1 for all albums...
i will contact qobuz service customer about that

**Comment by mherger** [comment]
I can confirm that at the beginning there would be higher load. But that's not unexpected, as I outlined. But this should mostly happen in the beginning. On my MacBook I did indeed see load of up to 35% for the first 2-3 seconds, but very little load thereafter. 

If this doesn't have a negative impact, then I'd just ignore it.

**Comment by th2j** [comment]
it is strange that for me, at the beginning of the playback, the load is 100% (with 2 or 3 cuts in the sound), and even after during playback, there are regular cpu spikes. I think this is abnormal. Also when i add some tracks to the playlist, again the cpu reachs 100% and the sound are cutting.

**Comment by th2j** [comment]
can you modify the qobuz playback from database to be as the playback from the qobuz plugin ?

**Comment by michaelherger** [comment]
Just use the Qobuz menus instead? You don't have hardly any local tracks anyway?

**Comment by michaelherger** [comment]
The advantage of the local menus is its better integration with the database. If you want to disable that better integration then there's little reason to use it.

**Comment by th2j** [comment]
yes the qobuz integration in the database is so much better, i can't go back,

but i still don't understand...
when i start a playlist, it is ok that lms works to query some infos at the begining,
but why it continues during all the playback... ?
it looks like lms queries some infos and don't find it and query again and again, etc

**Comment by michaelherger** [comment]
I can't say, I'm sorry. Because it's not what I see. Do you have the web UI open in your browser all the time, or are you using mobile apps or other clients? These could cause this polling.

**Comment by th2j** [comment]
ok you are right, i uninstalled the plugin material skin, and it is much better
